### ISSUE TITLE
CMake: PORTABLE=ON does not set HAVE_SSE on X86 builds => no fast CRC32

### ISSUE THREAD

**Issue by socketpair** [facebook/rocksdb#11311]
For example, in Fedora 37, RocksDB writes in logs:

`Fast CRC32 supported: Not supported on x86`

Fedora compiles RocksDB as PORTABLE=ON.

This patch reveals the problem:
```patch
--- a/util/crc32c.cc
+++ b/util/crc32c.cc
@@ -424,6 +424,7 @@
.
 static bool isSSE42() {
 #ifndef HAVE_SSE42
+#error here
   return false;
 #elif defined(__GNUC__) && defined(__x86_64__) && !defined(IOS_CROSS_COMPILE)
   uint32_t c_;
```

So, PORTABLE=ON must check if compiler supports SSE intrinsics, i.e., it MUST figure out if HAVE_SSE42 supported. And write fat big warning if not. (I would abort compilation if a compiler does not support).

Unfortunately, setting this options to true when PORTABLE=ON will possibly make invalid operation on old CPUs:

`util/math.h`:
```C++
#if defined(HAVE_SSE42) && (defined(_M_X64) || defined(_M_IX86))
    return static_cast<int>(__popcnt(static_cast<uint32_t>(v)));
#else
    return detail::BitsSetToOneFallback(static_cast<uint32_t>(v));
#endif
```

As you can see, it unconditionally (I mean at runtime) uses `popcnt`.

My suggestion -- is to rename the option to COMPILER_SUPPORTS_SSE42 + add runtime check for math functions.

The same for other options like AVX512 and so on.

Also, please define somewhere what means PORTABLE - "runs everywhere, but slow" or "Detect CPU feature at runtime".

**Comment by siying** [comment]
> Also, please define somewhere what means PORTABLE - "runs everywhere, but slow" or "Detect CPU feature at runtime".
This sounds good to me. Feel free to contribute a Pull Request and we will be happy to merge.

I would rely on @adamretter @mdcallag and @pdillinger for opinions on other suggestions on cross platform builds.

**Comment by socketpair** [comment]
@siying It's not about cross-platform build. I speak about building on x86_64 and then run on X86_64.

**Comment by siying** [comment]
> @siying It's not about cross-platform build. I speak about building on x86_64 and then run on X86_64.

Sure. No matter what we call it. We would rely on them to comment.

**Comment by mdcallag** [comment]
There have been struggles with this in the past in my builds, builds for Percona/MyRocks and also for MariaDB/MyRocks. A few details from me are at https://github.com/facebook/rocksdb/issues/2488

**Comment by socketpair** [comment]
possibly fixed by 55d58d91e7784a42e116f1055b5dee9fe077c021

in versions >= 7.9.2

**Comment by alanpaxton** [comment]
Hi @socketpair at @adamretter 's request I've spent some time looking into this.

I've been using a recent-ish main branch (50e9b3f9c7e9198257d0398bf14c8ddad70527cb, 5th March). I'm using a virtualised Ubuntu x86 which reports as GenuineIntel, Ivy Bridge..

Running
```
$ (mkdir build-portable; cd build-portable; cmake -DPORTABLE=ON ..; make -j8; ./db_basic_test)
```
runs the test, and is compiled with -DHAVE_SSE42, and reports in the log, as you observe: "Fast CRC32 supported: Not supported on x86"

It's exactly the same without portable:
```
$ (mkdir build-cmake; cd build-cmake; cmake  ..; make -j8; ./db_basic_test)
```

You can check how crc32c.cc is compiled if you just force recompilation of that:
```
$ (cd build-portable; touch ../util/crc32c.cc; make -j8 VERBOSE=1)
```

So the compile-time check in CMakeLists.txt must have succeeded for me:
```
CHECK_CXX_SOURCE_COMPILES("
#include <cstdint>
#include <nmmintrin.h>
#include <wmmintrin.h>
int main() {
  volatile uint32_t x = _mm_crc32_u32(0, 0);
  const auto a = _mm_set_epi64x(0, 0);
  const auto b = _mm_set_epi64x(0, 0);
  const auto c = _mm_clmulepi64_si128(a, b, 0x00);
  auto d = _mm_cvtsi128_si64(c);
}
" HAVE_SSE42)
```

Note that you can force the build to fail if it doesn't compile the above ,that is what -DFORCE_SSE42 will do. But I don't see how this is ever going to fail, when compiled with -msse4.2 and -mpclmul. It's at compile time ??

But understandably, at least for my machine, the runtime checks fail in `isSSE42()` which is why the "Fast CRC32 supported: Not supported on x86" message appears; neither the PCLMULQDQ nor SSE4.2 flags are set in the machine features when queried at runtime.

I put together I tiny test program to show more details
```
g++ -std=c++17 -DHAVE_SSE42 -msse4.2 -DHAVE_PCLMUL -mpclmul sse_test.cc
```
[sse_test.cc.zip](https://github.com/facebook/rocksdb/files/11077437/sse_test.cc.zip)

So what can or must we do ?
The debate on what to do about this seems to have gone round a few times - there's a comment in build_detect_platform from @pdillinger , re the Makefile-based build
  # The USE_SSE flag now means "attempt to compile with widely-available
  # Intel architecture extensions utilized by specific optimizations in the
  # source code." It's a qualifier on PORTABLE=1 that means "mostly portable."
  # It doesn't even really check that your current CPU is compatible.

1. I don't think the compile time checks are doing any good at all.
2. There must be plenty of building going on on virtualized machines that don't have/report the extensions at runtime.
3. Is it enough to build a non-portable build that sets  SSE42, PCLMUL et al and another lowest-common-denominator build that does not ? Do we need a build that does runtime checks ?

**Comment by socketpair** [comment]
I would suggest the following:

1. remove PORTABLE=1 (i.e. always build portable)
2. split CRC32 to different arch-specific .c files for CRC32 (same for math.h).
3. Compile arch-specific files using arch-specific flags. Let it break if a compiler does not support such flags. Actually, no real compiler that don't support these intrinsics.
4. detect at runtime what version to use and use :)

**Comment by mrambacher** [comment]
It looks like cpuid can be used to determine the feature set available on at least some hardware (x86, M1?) at run-time.  This might be a better solution than the compile-time flags (meaning option 5).

**Comment by adamretter** [comment]
@mrambacher Am I right in thinking that this an additional tool? This appears to not be part of the LSB, so would need to be installed additionally (at least on Ubuntu). Also it looks like you would need to `brew install cpuid` on macOS.

**Comment by mrambacher** [comment]
> @mrambacher Am I right in thinking that this an additional tool? This appears to not be part of the LSB, so would need to be installed additionally (at least on Ubuntu). Also it looks like you would need to `brew install cpuid` on macOS.

@adamretter I have not done all that much research yet but do not believe it is an additional tool, at least on X86 boxes.  For non-X86, I am not certain at this point.  I know there are toolkits that are built to provide these checks at run-time.

**Comment by ajkr** [comment]
> possibly fixed by [55d58d9](https://github.com/facebook/rocksdb/commit/55d58d91e7784a42e116f1055b5dee9fe077c021) (#10667)
> 
> in versions >= 7.9.2

Probably not. That relies on an MSVC-specific feature that compiles explicitly-used intrinsics to SSE4.2 instructions even when `-msse4.2` (or whatever the MSVC flag is called) is not provided. I didn't find an equivalent feature in gcc or anywhere else, which means we have to either (a) provide `-msse4.2` and let such instructions be emitted everywhere, e.g., string copying; or (b) not provide `-msse4.2` and then be unable to compile code guarded by `HAVE_SSE42`. For portable builds we do (b) so indeed it should not set HAVE_SSE.

**Comment by pdillinger** [comment]
> #11419 is my proposal to simplify everything.

That change landing should make things simpler. No more runtime detection craziness, deprecated USE_SSE. PORTABLE=1 means maximally portable, PORTABLE=0 means best for current CPU, and PORTABLE=foo means use "foo" as the target architecture (understood by your compiler).
